framework/data/ActiveDataProvider.php

```diff
@@ -93,14 +94,40 @@ public function init()
     }
 
     /**
-     * {@inheritdoc}
+     * Creates a wrapper of [[query]] that allows adding limit and order.
+     * @return QueryInterface
+     * @throws InvalidConfigException
      */
-    protected function prepareModels()
+    protected function createQueryWrapper(): QueryInterface
     {
         if (!$this->query instanceof QueryInterface) {
             throw new InvalidConfigException('The "query" property must be an instance of a class that implements the QueryInterface e.g. yii\db\Query or its subclasses.');
         }
-        $query = clone $this->query;
+        $wrapper = clone $this->query;
+        if ($wrapper instanceof Query && !empty($wrapper->union)) {
+            $wrapper->where = [];
+            $wrapper->limit = null;
+            $wrapper->offset = null;
+            $wrapper->orderBy = [];
+            $wrapper->selectOption = null;
+            $wrapper->distinct = false;
+            $wrapper->groupBy = [];
+            $wrapper->join = [];
+            $wrapper->having = [];
+            $wrapper->union = [];
+            $wrapper->params = [];
+            $wrapper->withQueries = [];
+            $wrapper->select('*')->from(['q' => $this->query]);
```

Comment:
There might be more properties that needs to be reset. For example `ActiveQuery` has `on` and `joinWith` (and `sql`?) properties, that may leak to wrapper. Developer may extend these classes and create own properties with the same problem. `Query` needs some kind of `reset()` method, that would be responsible for resting query state.

Or we cold use `new Query()`, similar to `ActiveQuery::queryScalar()`:
https://github.com/yiisoft/yii2/blob/b0b7832235a706625fca00221039227f15dde18f/framework/db/ActiveQuery.php#L355

Reply:
Using a `new Query()` will break `prepareModels()` itself if `query` is an `ActiveQuery` and the expected return value of `ActiveQuery::all()` is array of objects rather then an array of arrays. I'll think about the solution a little more.